src/main/scala/uk/gov/hmrc/play/audit/http/HttpAuditing.scala

```diff
@@ -84,10 +94,90 @@ trait HttpAuditing extends DateTimeUtils {
       hc.names.token -> hc.token.map(_.value).getOrElse("-"),
       Path -> request.url,
       Method -> request.verb) ++
-      request.body.map(b => Seq(RequestBody -> b.toString)).getOrElse(Seq.empty) ++
+      request.body.map(b => Seq(RequestBody -> maskRequestBody(b))).getOrElse(Seq.empty) ++
       HeaderFieldsExtractor.optionalAuditFields(hc.extraHeaders.toMap)
   }
 
+  private def maskRequestBody(body:Any):String = {
+    //The request body comes from calls to executeHooks in http-verbs
+    //It is either called with
+    // - a Map in the case of a web form
+    // - a String created from Json.stringify(wts.writes(... in the case of a class
+    // - a String in the case of a string (where the string can be XML)
+    body match {
+      case m: Map[_, _] => m.map {
+        case (key, _) if maskField(key.asInstanceOf[String]) => (key, MaskValue)
+        case other => other
+      }.toString
+      case s: String => maskString(s)
+      case other => throw new Exception(s"Unexpected type for requestBody when auditing ${outboundCallAuditType}: ${other.getClass}")
+    }
+  }
+
+  private def maskString(text:String) = {
+    if (text.startsWith("{")) {
+      try {
+        Json.stringify(maskJsonFields(Json.parse(text)))
+      } catch {
+        case e:JsonParseException => text
+      }
+    } else if (text.startsWith("<")) {
+      try {
+        val builder = new StringBuilder
+        PrettyPrinter.format(maskXMLFields(xxeResistantParser.loadString(text)), builder)
+        builder.toString()
+      } catch {
+        case e: SAXParseException => text
+      }
+    } else {
+      text
+    }
+  }
+
+  private def maskJsonFields(json: JsValue): JsValue = json match {
+    case JsObject(fields) => JsObject(fields.map {
+      case (key,_) if maskField(key) => (key,JsString(MaskValue))
+      case (key,value) => (key,maskJsonFields(value))
+    })
+    case JsArray(values) => JsArray(values.map(v => maskJsonFields(v)))
+    case other => other
+  }
+
+  private def maskXMLFields(elem: Elem):Elem = {
+    elem.copy(
+      child = elem.child.map { maskXMLFields(_) },
+      attributes = maskXMLAttributes(elem.attributes)
+    )
+  }
+
+  private def maskXMLFields(node:Node):Node = {
+    node match {
+      case e:Elem if maskField(e.label) => e.copy(child = Seq(Text(MaskValue)), attributes = maskXMLAttributes(e.attributes))
+      case e:Elem => e.copy(child = e.child.map(maskXMLFields(_)), attributes = maskXMLAttributes(e.attributes))
+      case other => other
+    }
+  }
+
+  private def maskXMLAttributes(attributes: MetaData):MetaData = {
+    attributes.foldLeft(Null: scala.xml.MetaData) { (current,attr) =>
```

Comment:
it may be a personal preference but `current` is the accumulated value here but I would read it as current element being processed and not the one that was processed before